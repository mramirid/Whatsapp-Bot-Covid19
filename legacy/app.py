from flask import Flask, render_template, request, url_for, redirect
from flask_mysqldb import MySQL
from twilio.twiml.messaging_response import MessagingResponse
from datetime import datetime

import functions.getter as getter
import json

app = Flask(__name__)
app.config['MYSQL_HOST'] = 'localhost'
app.config['MYSQL_USER'] = 'root'
app.config['MYSQL_PASSWORD'] = ''
app.config['MYSQL_DB'] = 'whatsapp_bot_covid'
mysql = MySQL(app)

# init mysql connection
getter.init_connection(mysql)


@app.route('/')
def base():
    return 'LANJUTKAN'


def myconverter(o):
    if isinstance(o, datetime):
        return o.__str__()


@app.route('/nasional')
def test_nasional():
    data = getter.get_nasional()
    datas = json.dumps(data, indent=4, sort_keys=True, default=myconverter)
    return str(datas)


@app.route('/provinsi')
def test_provinsi():
    data = getter.get_prov_byname('jawa timur')
    datas = json.dumps(data, indent=4, sort_keys=True, default=myconverter)
    return str(datas)


@app.route('/chat', methods=['POST'])
def sms_reply():
    # Ambil pesan dari chat
    message = request.form.get('Body').lower().strip()

    # Jika pesan bukan command, abaikan
    if message[0] != '/':
        return

    response = ''
    chat_response = MessagingResponse()
    words = message.split()

    if words[0] == '/halo':
        response = "Halo juga\n\n"
        response += "Saya adalah bot yang didesain untuk memberikan info kasus COVID-19 baik secara nasional maupun berdasarkan provinsi.\n"
        response += "Sumber data didapatkan dari API yang disediakan oleh kawalcorona - https://kawalcorona.com/api/\n"
        response += "Ketikkan '/help' untuk mendapatkan informasi mengenai fitur yang diprogram untuk saya\n\n"
    elif words[0] == '/nasional':
        response = getter.get_nasional()
    elif words[0] == '/cari':
        if len(words) > 1:
            nama_prov = ' '.join(words[1:])
            result = getter.get_prov_byname(nama_prov)
            response = result if result else 'Provinsi tidak ditemukan'
        else:
            response += 'Pastikan anda sudah memasukan nama provinsi'
    elif words[0] == '/help':
        response = 'List command:\n\n'
        response += '1. /halo -> Perkenalan bot\n'
        response += '2. /nasional -> Kasus COVID-19 di Indonesia\n'
        response += '3. /cari [nama_provinsi] -> Mencari kasus COVID-19 berdasarkan provinsi\n'
        response += 'Misal: /cari jawa timur\n\n'
        response += 'Versi 1.0 - 17/05/2020 20.00'
    else:
        response = 'Maaf perintah tidak dikenali, ketik /help untuk mengetahui fitur'

    chat_response.message(response)

    return str(chat_response)


# Ubah menjadi debug=False ketika akan dideploy ke hosting
if __name__ == "__main__":
    app.run(debug=True)
