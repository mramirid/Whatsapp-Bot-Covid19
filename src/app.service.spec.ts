import { of } from 'rxjs';
import { AppService } from './app.service';
import type { CountryService } from './country/country.service';

const getStatsMessageSpy = jest.fn(() => of('COUNTRY_STATS_MESSAGE'));
const countryServiceMock: Partial<CountryService> = {
  getStatsMessage: getStatsMessageSpy,
};
const appService = new AppService(countryServiceMock as any);

describe('replyHello()', () => {
  it('should reply hello correctly', () => {
    expect(appService.replyHello('Amir')).toMatchSnapshot();
  });
});

describe('replyHelp()', () => {
  it('should reply help correctly', () => {
    expect(appService.replyHelp()).toMatchSnapshot();
  });
});

describe('replyUnknown()', () => {
  it('should reply unknown correctly', () => {
    expect(appService.replyUnknown()).toMatchSnapshot();
  });
});

describe('replyNasional()', () => {
  it('should call CountryService.getStatsMessage() in order to reply nasional', (done) => {
    appService.replyNasional().subscribe({
      next: () => {
        expect(getStatsMessageSpy).toBeCalled();
      },
      error: done,
      complete: done,
    });
  });
});
