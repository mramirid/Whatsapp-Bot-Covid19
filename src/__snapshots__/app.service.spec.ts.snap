// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replyHello() should reply hello correctly 1`] = `
"Halo Amir,

Saya adalah bot yang didesain untuk memberikan info kasus COVID-19 di Indonesia.
Sumber data didapatkan dari API yang disediakan oleh Open Disease Data API (https://corona.lmao.ninja/).
Ketikkan \`\`\`/help\`\`\` untuk mendapatkan informasi mengenai saya dan daftar perintah yang tersedia."
`;

exports[`replyHelp() should reply help correctly 1`] = `
"Daftar perintah:
1. \`\`\`/halo\`\`\` -> Perkenalan bot
2. \`\`\`/nasional\`\`\` -> Statistik COVID-19 di Indonesia
3. \`\`\`/help\`\`\` -> Daftar perintah

_Versi 1.0 - 17/05/2020 20.00: Bot dibuat._
_Versi 2.0 - 13/05/2023 10.06: Penggantian API & penghapusan fitur statistik per-provinsi._"
`;

exports[`replyUnknown() should reply unknown correctly 1`] = `"Maaf perintah tidak dikenali, kirimkan saya \`\`\`/help\`\`\` untuk melihat daftar perintah."`;
